### DISCUSSION TITLE
working-with-the-result-pattern-part-2-safety-and-simplicity-with-linq/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#438] [Announcements]
# working-with-the-result-pattern-part-2-safety-and-simplicity-with-linq/

In this post I show how to clean up unreadable Switch()-based result pattern code using LINQ query syntax, giving both readability and safety

https://andrewlock.net/working-with-the-result-pattern-part-2-safety-and-simplicity-with-linq/

**Comment by thomaslevesque** [comment]
Interesting. I have actually developed an Option<T> [library](https://github.com/thomaslevesque/Hamlet?tab=readme-ov-file#usage-with-linq) that uses this Linq trick, but never realized it could also be applied to Result<T>

**Comment by dcarapic** [comment]
I find these kind of articles interesting as some sort of mind exercises but (IMHO) I think that the Result/Option patterns are just not a good fit for c#. While you can (mis)use the LINQ to chain the operators and it might look 'nice' (which is subjective) it is far removed from idiomatic c#. If I were to scroll to the middle of the article and look at the LINQ 'from' statements I would have assumed it was some sort of nested collection data extraction and not a set of imperative statements. Should I store intermediate values via 'let' clause then ... go full LINQ? At that point why even bother with c#, switch to f# and be done with it.
I do like the discriminating unions in languages such as Typescript, but there they are just natural to use.
Potentially the new discriminating union proposal will make such things more accessible in c#.

**Comment by MarsonShine** [comment]
question:
`Select`/`SelectMany` method: `onFailure: e => Result<TResult>.Fail(e)` was wrong: Result<TResult> cant convert to 
Or is this meant to introduce the third article in the series?

**Reply by MasterMann** [reply]
Was wondering if anyone else had the same issue.

**Comment by jonatwabash** [comment]
Honestly, this `Result<T>` approach seems to me like it's a technical solution to what is generally not a technical problem.

I mean, depending on how the actual business works, the original code in the previous post might be valid as is, at least if the business has a default tenant ID to assign users to, and if that's not the situation, the easiest fix may be to make everything nullable and have each function immediately return `null` if the relevant input is null or empty. This would very much obscure what went wrong when things go wrong, but logging is generally a much more effective way to track that sort of information than exception handling assuming the devil team puts in the time and energy to understand what needs to be logged and when.

Simultaneously, how much information to report out to app users, especially when something goes wrong, is a significant business question, especially when dealing with user or customer identifying data.

I can see a more of a case for unions more generally, which presumably could be used to set up these sorts of things, but I'd expect unions to frequently have rather more types included than just T and error.

**Reply by repodn** [reply]
This is getting a little far-fetched.

**Comment by drdamour** [comment]
now the c# is looking more like scala with its map and flatMap on it's "single" containers like Option and Future and probably some day (sadly) Result[T,E]
